### ISSUE TITLE
[Bug] Hot Reload crashes when resource in App.xaml changes

### ISSUE THREAD

**Issue by viktorszekeress** [xamarin/Xamarin.Forms#9309]
### Description
I experience a crash when `Shell` is used and I change a resource defined in `App.xaml`. I have 3 pages (`FlyoutItem`s) defined in `Shell`, each page uses label style defined in `App.xaml`. If I navigate to 1st or 2nd page and change the style, crash occurs. Crash doesn't occur when I'm navigated to 3rd (last) page and I change the style.

### Steps to Reproduce

1. Open attached solution and run the app
2. Navigate to Settings page (2nd page)
3. Change color defined in `TextStyle` in `App.xaml` and save the file to trigger Hot Reload
4. Crash occurs

### Expected Behavior
No crash occurring after updating the style.

### Actual Behavior
App crashes with exception:
```
System.ArgumentException: 'Ambiguous routes matched for: //IMPL_about/IMPL_about/about/ matches found: //IMPL_about/IMPL_about/about,//IMPL_about/IMPL_about/about
Parameter name: uri'
```

### Basic Information

- Version with issue: 4.4.0.991477

### Screenshots

![settings_page](https://user-images.githubusercontent.com/45604126/73057091-e9fb6180-3e90-11ea-8ca0-c47330df263e.jpg)
![app xaml](https://user-images.githubusercontent.com/45604126/73057097-ed8ee880-3e90-11ea-9c35-fe93e6593e18.jpg)
![hot_reload_crash](https://user-images.githubusercontent.com/45604126/73057100-f1226f80-3e90-11ea-8f88-bdc1fdd628c1.jpg)

### Reproduction Link

[HotReloadResourceCrash.zip](https://github.com/xamarin/Xamarin.Forms/files/4107543/HotReloadResourceCrash.zip)

> VS bug [#1079452](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1079452)

**Comment by ChaseFlorell** [comment]
I can confirm I'm seeing this as well

**Comment by viktorszekeress** [comment]
This issue also occurs when any of my pages includes resource dictionary from file (via MergedDictionaries). If Settings page includes SomeResources.xaml and I change a resource in SomeResources.xaml, I get the same crash.
Thus, the only workaround is having all resources declared in all pages.

Can somebody take a look at this issue?

**Comment by ChaseFlorell** [comment]
@viktorszekeress good call on the Merged Dictionaries, I've got that too when I see the crash.

```xaml
<?xml version="1.0" encoding="UTF-8"?>
<Application xmlns="http://xamarin.com/schemas/2014/forms" 
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             ...>
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <xaml:Colors />
                <xaml:GlobalStyles />
                <xaml:TemplateStyles />
                <xaml:Fonts />
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Application.Resources>
</Application>
```

**Comment by GuidoNeele** [comment]
This is a problem with hot reload. Routes are registered twice when the shell XAML file is edited. After restarting the app everything is normal.

Xamarin Forms 4.5

**Comment by drasticactions** [comment]
Hey folks, Hot Reload dev here 👋

@GuidoNeele @viktorszekeress 

The issue you're seeing here is, as far as I can tell, is directly to do with Shell itself being invoked via Hot Reload. The resource dictionaries, however they're set up, don't matter since the Shell routes are being reinvoked, causing the `Ambiguous routes matched for` error to be thrown. You can invoke the same error editing any of the lines inside of the `AppShell.xaml` too, and you can reproduce this in the [Xaminals](https://github.com/xamarin/xamarin-forms-samples/blob/master/UserInterface/Xaminals/Xaminals/AppShell.xaml) shell sample.

Generally, in the current implementation of XAML Hot Reload, when you make a change to the `App.xaml`, we only invoke `[XamlLoader.LoadResources](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Xaml/XamlLoader.cs#L158)`, which replaces the loaded resources in your app. For other changes (RDs, ContentPages, etc), is uses `[XamlLoader.Load](https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Xaml/XamlLoader.cs#L78)`

In the case of a Shell app, if you touch elements that end up poking at the `AppShell`, it ends up reloading the `AppShell`, which then causes the routes to be reapplied, causing the error you see.

Now, we (Hot Reload) are working on more enhance editing of changes that could avoid the issue you see here (since you're not changing anything directly in Shell, it shouldn't be trying to reinvoke the routes). However, the underlying issue in Forms still exists. If you actually _were_ trying to edit the routes on your page, the code handling routes would be reapplied, causing the bug to appear again.

@ChaseFlorell The issue you're seeing is unrelated to this one: From what I can tell, there seems to be a bug with how Xamarin.Forms handles the virtualization of the CollectionView, hence it poking elements that are disposed of. Again, this feels more like an issue invoked via Hot Reload, but not something that can solely be fixed _within_ Hot Reload. I would file that separately if it's not already.

**Comment by viktorszekeress** [comment]
Hi @drasticactions, thanks for comprehensive explanation!

From my understanding:
You are now working on fix that will solve the issue when changing resources in App.xaml crashes the app, by avoiding re-invoking the routes in Shell.xaml.

What about the case when I change resources in Shell.xaml? I don't have such scenario with my app, but I can imagine other devs might have it.

Another question - how about my case about merged dictionaries (referenced in pages as MergedDictionaries), fits in the bug description? Does changing resources in dictionary referenced as MergedDictionary invoke routes as well?

**Comment by drasticactions** [comment]
@viktorszekeress 

**TL;DR** The actual bug is with Xamarin.Forms Shell not being able to handle reapplying Shell routes. It's invoked via Hot Reload, but the underlying issue is with Forms, so there's nothing for me to fix.

I _am_ working on more incremental ways of editing the XAML Resources that may sidestep bugs like this one, but the actual underlying bug (Reapplying routes on reloads) needs to be addressed.
---

The issue here, as far as I can tell, is specific to using Xamarin.Forms shell and having assets that end up touching `AppShell`, causing it to reapply routes. The other issues you noted (Crashing when editing App.xaml, Merged Dictionaries) are _all related_ to this underlying Shell issue: If you're seeing `Ambiguous routes matched for` being thrown on a change, that is this bug, no matter what it's coming from, since it's calling the same code to reapply the routes. 

If you try to replicate it in a non-shell app, they (should) work fine. It's because this is a Shell app that's causing this issue. If you can reproduce that in a _non-shell_ app, then that's something else for me to check out.

The Forms code, as far as I can see, only assumes that the routes will be set once, and never again. That, I believe, needs to be changed so that it can be altered and updated if a user does a Hot Reload.

**Comment by viktorszekeress** [comment]
That clears thing up. Thanks!

**Comment by PureWeen** [comment]
This is fixed on the latest version of Forms

I tested the original reproduction and it works fine when updated to XF 4.6
